components/script/dom/notification.rs

```diff
@@ -0,0 +1,348 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::cell::Cell;
+
+use dom_struct::dom_struct;
+use js::jsapi::Heap;
+use js::jsval::JSVal;
+use js::rust::{HandleObject, MutableHandleValue};
+use num_traits::ToPrimitive;
+use servo_url::{ImmutableOrigin, ServoUrl};
+
+use super::bindings::codegen::Bindings::NotificationBinding::{
+    NotificationAction, NotificationOptions,
+};
+use super::bindings::codegen::Bindings::PerformanceBinding::PerformanceMethods;
+use super::bindings::reflector::reflect_dom_object_with_proto;
+use super::bindings::str::USVString;
+use crate::dom::bindings::codegen::Bindings::NotificationBinding::{
+    NotificationDirection, NotificationMethods, NotificationPermission,
+    NotificationPermissionCallback,
+};
+use crate::dom::bindings::codegen::Bindings::PermissionStatusBinding::{
+    PermissionName, PermissionState,
+};
+use crate::dom::bindings::error::Error;
+use crate::dom::bindings::import::module::{Fallible, Rc};
+use crate::dom::bindings::inheritance::Castable;
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::bindings::trace::RootedTraceableBox;
+use crate::dom::bindings::utils::to_frozen_array;
+use crate::dom::eventtarget::EventTarget;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::permissions::get_descriptor_permission_state;
+use crate::dom::promise::Promise;
+use crate::dom::serviceworkerglobalscope::ServiceWorkerGlobalScope;
+use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
+
+/// <https://notifications.spec.whatwg.org/#api>
+#[dom_struct]
+pub struct Notification {
+    eventtarget: EventTarget,
+    title: DOMString,
+    body: DOMString,
+    #[ignore_malloc_size_of = "mozjs"]
+    data: Box<Heap<JSVal>>,
+    dir: NotificationDirection,
+    image: Option<USVString>,
+    icon: Option<USVString>,
+    badge: Option<USVString>,
+    lang: DOMString,
+    silent: Option<bool>,
+    tag: DOMString,
+    #[no_trace] // ImmutableOrigin is not traceable
```

Comment:
Instead of storing all these things, why not just store the options?

Reply:
Some of these are the same as the value in the options, but others require to be processed, transformed, or given a fallback value before we store it.